### ISSUE TITLE
how to self-host an instance of cors-anywhere on firebase function

### ISSUE THREAD

**Issue by nathann282** [Rob--W/cors-anywhere#160]
I tried to run one on heroku and it works perfectly but now I'm planning to move from heroku to firebase. And got stuck with how to implement this on firebase function

**Comment by Rob--W** [comment]
Please stop with opening new tickets for the same question. You can post comments at your previous ticket, even after it was closed (https://github.com/Rob--W/cors-anywhere/issues/158).

As this is merely a question and not specific to my library, let alone requiring actions on my part, I'll close it again.

Since your question already mentions "firebase function", I assume that you know how to get a Node.js application to run on Firebase, and I will only briefly explain how to hook up CORS Anywhere with a Cloud Function at Firebase. I haven't used Firebase for hosting before, so the following has not been tested, but should at least give a hint on how you can use CORS Anywhere on Firebase.

The relevant documentation is at:

- https://firebase.google.com/docs/hosting/
- https://firebase.google.com/docs/hosting/functions
- https://firebase.google.com/docs/functions/http-events

The `onRequest` handler appears to pass an object that is compatible with the Express.js API. There has already been a previous question about using CORS Anywhere with Express.js, see https://github.com/Rob--W/cors-anywhere/issues/81 for more references.
Basically, receive the request from Firebase and emit the request to CORS Anywhere, as a server library; for the documentation, see https://github.com/Rob--W/cors-anywhere#server .

```javascript
var cors_proxy = require('cors-anywhere').createServer({
  requireHeader: ['origin', 'x-requested-with'],
  removeHeaders: [
    'cookie',
    'cookie2',
  ],
  // See README.md for other options
});

var functions = require('firebase-functions');
exports.proxyWithCorsAnywhere = functions.https.onRequest(req, res) {
  cors_proxy.emit('request', req, res);
};
```

`firebase.json` :
```json
{
 "hosting": {
   // ...

   // Add the "rewrites" attribute within "hosting"
   "rewrites": [ {
     "source": "**",
     "function": "proxyWithCorsAnywhere"
   } ]
 }
}
```

**Comment by nuthinking** [comment]
Is this still supposed to work? I can't make it work on my end following exactly these and other instructions:

<img width="877" alt="Screenshot 2022-04-12 at 14 40 06" src="https://user-images.githubusercontent.com/163635/162964442-5dec2c22-35e0-4bf8-8e51-eaa5d9307706.png">

**Comment by brumasribera** [comment]
Thank you @Rob--W for the lib and the help here! 

I made it run by adding the lib on the dependencies in the package.json of the /functions folder:

```
"dependencies": {
    "cors-anywhere": "^0.4.4",
    "firebase-admin": "^10.0.2",
    "firebase-functions": "^3.18.0"
  },
```

I got the following error though where trying to use it following, for example this url: 
http://localhost:5000/http://google.com/

I get the following error:
`Missing required request header. Must specify one of: origin,x-requested-with`

As a bonus, it's important to know that in order to run functions you need a 'pay-as-you-go' plan, so unfortunately is not free as with Heroku.

**Comment by isoaxe** [comment]
I've managed to successfully run a locally hosted `cors-anywhere` server using Firebase functions. You can find the repo [here](https://github.com/Isoaxe/cors-server). However, when using the hosted instance, I get the same old errors in the Firefox browser console:

<img width="650" alt="image" src="https://user-images.githubusercontent.com/63400356/188560612-28437377-bd54-4ccd-8bd2-76f9cec3a5f1.png">

And a similar one in Chrome:

<img width="500" alt="image" src="https://user-images.githubusercontent.com/63400356/188562624-856f6997-ac38-4c93-bef9-6534259d4703.png">

Navigating to the hosted proxy server [url](https://us-central1-private-cors-server.cloudfunctions.net/proxy) gives the expected help page as with the local instance.

Any ideas on what the issue might be and how to rectify it?

**Comment by isoaxe** [comment]
Did anyone get a server working when hosted remote and using Firebase functions? Still can't figure it out.

**Comment by weilinzung** [comment]
> I've managed to successfully run a locally hosted `cors-anywhere` server using Firebase functions. You can find the repo [here](https://github.com/Isoaxe/cors-server). However, when using the hosted instance, I get the same old errors in the Firefox browser console:
> 
> <img alt="image" width="650" src="https://user-images.githubusercontent.com/63400356/188560612-28437377-bd54-4ccd-8bd2-76f9cec3a5f1.png">
> 
> And a similar one in Chrome:
> 
> <img alt="image" width="500" src="https://user-images.githubusercontent.com/63400356/188562624-856f6997-ac38-4c93-bef9-6534259d4703.png">
> 
> Navigating to the hosted proxy server [url](https://us-central1-private-cors-server.cloudfunctions.net/proxy) gives the expected help page as with the local instance.
> 
> Any ideas on what the issue might be and how to rectify it?

Hi @Isoaxe, have you tried to enable cors with lib for Firebase function? 

https://stackoverflow.com/a/42756623 or
 https://stackoverflow.com/a/50992145

**Comment by isaackoz** [comment]
> I've managed to successfully run a locally hosted `cors-anywhere` server using Firebase functions. You can find the repo [here](https://github.com/Isoaxe/cors-server). However, when using the hosted instance, I get the same old errors in the Firefox browser console:
> 
> <img alt="image" width="650" src="https://user-images.githubusercontent.com/63400356/188560612-28437377-bd54-4ccd-8bd2-76f9cec3a5f1.png">
> 
> And a similar one in Chrome:
> 
> <img alt="image" width="500" src="https://user-images.githubusercontent.com/63400356/188562624-856f6997-ac38-4c93-bef9-6534259d4703.png">
> 
> Navigating to the hosted proxy server [url](https://us-central1-private-cors-server.cloudfunctions.net/proxy) gives the expected help page as with the local instance.
> 
> Any ideas on what the issue might be and how to rectify it?

I made a merge request fixing this issue thanks to @weilinzung's advice.

**Comment by nuthinking** [comment]
@isaackoz where is the PR?

**Comment by isaackoz** [comment]
> @isaackoz where is the PR?

Apologies, I should have specified and included the fix here. It was a PR for @Isoaxe's repo found [here](https://github.com/Isoaxe/cors-server/pull/3). 

```const {onRequest} = require("firebase-functions/v2/https");
const corsAnywhere = require('cors-anywhere');
const cors = require("cors")({origin:true})
const corsServer = corsAnywhere.createServer({
    originWhitelist: [
      'http://localhost:3000',
    ],
    requireHeader: ['origin', 'x-requested-with'],
    removeHeaders: ['cookie', 'cookie2'],
});

exports.proxy = onRequest((request, response) => {
  cors(request,response,() =>{
    corsServer.emit('request', request, response);
  })    
});
```
The main fix was specifying cors and setting the origin to true. You can also specify a specific origin here as well. True allows all origins. `const cors = require("cors")({origin:true})`

**Comment by isoaxe** [comment]
Hi guys, just back from holidays so hadn't got a chance to look at any of that yet.

@weilinzung @isaackoz Thanks for your input. I'll take a look at the PR today at some point.

**Comment by nuthinking** [comment]
@isaackoz ah! I can see the changes. Thanks for sharing!

**Comment by isoaxe** [comment]
@isaackoz I reviewed the changes but have implemented seperately myself as there were several changes that I needed reversed for my own use of the server (including the `originWhitelist` array etc). Will merge the changes I made now.

**Comment by jaredsburrows** [comment]
Using @isoaxe example, I am still seeing `Missing required request header. Must specify one of: origin,x-requested-with`. Has anyone resolved this?

**Comment by isoaxe** [comment]
Hi @jaredsburrows, if you open an issue in [my repo](https://github.com/isoaxe/cors-server) I'll try and help.

**Comment by camilohr18** [comment]
hi @isoaxe I am having problems with the time to respond. I got a 504 error I left an issue on your repo.
